### DISCUSSION TITLE
CMake error when generating project for UWP using VS2022 17.8.0

### DISCUSSION THREAD

**Discussion by rh101** [axmolengine/axmol#1443] [Q&A] [answered]
- axmol version:
- devices test on:
- developing environments
   - NDK version: r23c
   - Xcode version: 14.2
   - Visual Studio: 
      - VS version: 2022 (17.8.0)
      - MSVC version: 1929, 1934, 19.35, 19.36, 19.37
      - Windows SDK version: 10.0.22621.0
   - cmake version: 3.28-rc5
Steps to Reproduce:

Run this on a new project:
`axmol build -p winuwp -c`

Results in this output:
```
build1k: PowerShell 7.3.9 on Microsoft Windows 10.0.19045.0
build1k:
Name                           Value
----                           -----
xb                             {--target, test}
xt                             cmake
sdk
cc
p                              winuwp
d                              G:\dev\test
prefix                         G:\game_engines\axmol\tools\external
a                              x64
dll                            False
xc                             {}


build1k: proj_dir=G:\dev\test, external_prefix=G:\game_engines\axmol\tools\external
build1k: Using glslcc: G:\game_engines\axmol\tools\external\glslcc\glslcc.exe, version: 1.9.3
build1k: Using cmake: C:\Program Files\CMake\bin\cmake.exe, version: 3.28.0
build1k: Using nuget: G:\game_engines\axmol\tools\external\nuget\nuget.exe, version: 5.5.1
build1k: Building target winuwp on windows with toolchain msvc ...
build1k: CONFIG_ALL_OPTIONS=-DCMAKE_SYSTEM_NAME=WindowsStore -DCMAKE_SYSTEM_VERSION=10.0 -A x64 -DCMAKE_INSTALL_PREFIX=G:\dev\test\install_winuwp_x64, Count=5
-- Building for: Visual Studio 17 2022
-- Selecting Windows SDK version 10.0.22621.0 to target Windows 10.0.
-- The C compiler identification is unknown
-- The CXX compiler identification is MSVC 19.38.33130.0
CMake Error at CMakeLists.txt:30 (project):
  No CMAKE_C_COMPILER could be found.



-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.38.33130/bin/Hostx64/x64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Configuring incomplete, errors occurred!
build1k: Generate done, press any key to continue . . .
```

Is there something in the cmake scripts that would cause this issue?  I reported it earlier to Microsoft [here](https://developercommunity.visualstudio.com/t/No-CMAKE_C_COMPILER-could-be-found-with-/10516801), assuming it was an issue with Visual Studio 2022 17.8.0, but they cannot reproduce it.

**Comment by halx99** [comment]
it's should be cmake bug, or vs bug

**Comment by halx99** [comment]
just test a very simple demo:

```cmake
cmake_minimum_required(VERSION 3.27.0)

add_library(dummy dummy.cpp)
```

also has error: 
```txt
build1k: CONFIG_ALL_OPTIONS=-DCMAKE_SYSTEM_NAME=WindowsStore -DCMAKE_SYSTEM_VERSION=10.0 -A x64 -DCMAKE_INSTALL_PREFIX=E:\dev\oss\test\install_winuwp_x64, Count=5
CMake Warning (dev) in CMakeLists.txt:
  No project() command is present.  The top-level CMakeLists.txt file must
  contain a literal, direct call to the project() command.  Add a line of
-- Building for: Visual Studio 17 2022
  code such as

    project(ProjectName)

  near the top of the file, but after cmake_minimum_required().

  CMake is pretending there is a "project(Project)" command on the first
  line.
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) in CMakeLists.txt:
  cmake_minimum_required() should be called prior to this top-level project()
  call.  Please see the cmake-commands(7) manual for usage documentation of
  both commands.
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Selecting Windows SDK version 10.0.22621.0 to target Windows 10.0.
-- The C compiler identification is unknown
-- The CXX compiler identification is MSVC 19.38.33130.0
CMake Error in CMakeLists.txt:
  No CMAKE_C_COMPILER could be found.



-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files/Microsoft Visual Studio/2022/Professional/VC/Tools/MSVC/14.38.33130/bin/Hostx64/x64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Configuring incomplete, errors occurred!
```

**Comment by rh101** [comment]
Thanks for testing that out.  I'm just a little confused as to how MS can't reproduce it.  I've posted step by step instructions on the MS site regarding how to reproduce this issue, so hopefully we can get clarification if it is a VS or cmake issue.  The post is flagged as private so may not be visible in that thread.

**Comment by halx99** [comment]
error occurred in cmake detect c compiler step.

vs2022 17.8.0 not support ForcedUsingFiles for c source files, it's support c++, so the cmake detect template vcxproj compile c source file failed

<img width="583" alt="image" src="https://github.com/axmolengine/axmol/assets/6977319/add7fa44-56fe-4cdc-a5aa-c6d37a69891e">

And the options was inherited from visual studio .props template, not settting by cmake, so this issue should a visual studio bug

error output:

```txt
1>------ Build started: Project: CompilerIdC, Configuration: Debug Win32 ------
1>C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppBuild.targets(519,5): warning MSB8021: The value 'MultiByte' of the variable 'CharacterSet' is incompatible with the value 'true' of the variable 'WindowsAppContainer'.
1>CMakeCCompilerId.c
1>/CMakeCCompilerId.c(1,1): error C1007: unrecognized flag '-FU' in 'c1'
1>(compiling source file '/CMakeCCompilerId.c')
```

**Comment by halx99** [comment]
There are 3 files define default compiler option "/FU":

- C:\Program Files\Microsoft Visual Studio\2022\Community\Msbuild\Microsoft\VC\v170\Microsoft.Cpp.AppContainerApplication.props
- C:\Program Files\Microsoft Visual Studio\2022\Community\Msbuild\Microsoft\VC\v160\Microsoft.Cpp.AppContainerApplication.props
- C:\Program Files\Microsoft Visual Studio\2022\Community\Msbuild\Microsoft\VC\v150\Microsoft.Cpp.Common.props

**Reply by halx99** [reply]
A workaround solution may work is modify `C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.Cpp.AppContainerApplication.props`
```xml
<ForcedUsingFiles>$(VCToolsInstallDir)lib\x86\store\references\platform.winmd;%(ForcedUsingFiles)</ForcedUsingFiles>
```

to

```xml
<ForcedUsingFiles Condition="'$(VCToolsVersion)' &lt; '14.38'">$(VCToolsInstallDir)lib\x86\store\references\platform.winmd;%(ForcedUsingFiles)</ForcedUsingFiles>
```

**Reply by rh101** [reply]
Great workaround! Tested it and cmake runs successfully for UWP.

**Comment by rh101** [comment]
Is that what is causing the issue with UWP project generation with CMake? If so, could you possibly add all of this info to [this post](https://developercommunity.visualstudio.com/t/No-CMAKE_C_COMPILER-could-be-found-with-/10516801) so they can hopefully fix it quicker?

**Comment by halx99** [comment, marked as answer]
The root cause maybe the vs2022 17.8 remove C compiler flag `FU` support.

**Comment by rh101** [comment]
Fair enough.  I see many logged issues regarding the "/FU" switch in v17.8 and the affect it is having on UWP projects, so it seems like they already know about it.

**Reply by halx99** [reply]
A cmake patch also was created: https://gitlab.kitware.com/cmake/cmake/-/merge_requests/9014 to solve this problem if @microsoft no plan to  fix it in vs17.8.x releases.

**Reply by rh101** [reply]
Excellent! I just checked the link you posted, and there is a response from Microsoft in there indicating that they will be releasing a fix for the issue next week.
